apps/oxlint/src-js/plugins/source_code.ts

```diff
@@ -169,29 +169,66 @@ export const SOURCE_CODE = Object.freeze({
    * @param nodeOrToken - The AST node or token to check for adjacent comment tokens.
    * @returns Array of `Comment`s in occurrence order.
    */
-  // oxlint-disable-next-line no-unused-vars
   getCommentsBefore(nodeOrToken: NodeOrToken): Comment[] {
-    throw new Error('`sourceCode.getCommentsBefore` not implemented yet'); // TODO
+    if (ast === null) initAst();
+    const { comments } = ast;
+    if (comments.length === 0) return [];
+    const targetStart = nodeOrToken.range[0];
+    let i = comments.length - 1;
+    // linear backwards search for the index of the last comment before the provided node or token, TODO(perf): binary
```

Comment:
Just a nit. Try to keep comments short where you can, and capitalize (I'm a capitalist!)

```suggestion
    // Linear backwards search for last comment before `nodeOrToken`.
    // TODO: Use binary search.
```